core/roboconf-dm/src/main/java/net/roboconf/dm/internal/api/impl/RandomMngrImpl.java

```diff
@@ -0,0 +1,292 @@
+
+		for( Instance instance : InstanceHelpers.getAllInstances( application )) {
+			generateRandomValues( application, instance );
+		}
+	}
+
+
+	@Override
+	public synchronized void releaseRandomValues( Application application, Instance instance ) {
+
+		// Only used for random ports.
+		// We only remove the ports used by this instance.
+		List<Integer> portsToRelease = new ArrayList<> ();
+		for( ExportedVariable var : instance.getComponent().exportedVariables.values()) {
+			if( ! var.isRandom()
+					|| var.getRandomKind() != RandomKind.PORT )
+				continue;
+
+			String value = instance.overriddenExports.get( var.getName());
+			if( value != null )
+				portsToRelease.add( Integer.parseInt( value ));
+		}
+
+		// Update the cache
+		InstanceContext ctx = findAgentContext( application, instance );
+		List<Integer> ports = this.agentToRandomPorts.get( ctx );
+		if( ports != null ) {
+			ports.removeAll( portsToRelease );
+			if( ports.isEmpty())
+				this.agentToRandomPorts.remove( ctx );
+		}
+	}
+
+
+	@Override
+	public void releaseAllRandomValues( Application application ) {
+
+		// Only used for random ports.
+		// Find the contexts to remove
+		List<InstanceContext> toRemove = new ArrayList<> ();
+		for( InstanceContext ctx : this.agentToRandomPorts.keySet()) {
+			if( ctx.getQualifier() == null
+					&& Objects.equals( application.getName(), ctx.getName()))
+				toRemove.add( ctx );
+		}
+
+		// Remove them
+		for( InstanceContext ctx : toRemove )
+			this.agentToRandomPorts.remove( ctx );
+	}
+
+
+	@Override
+	public void restoreRandomValuesCache( Application application ) {
+
+		// Only used for random ports.
+		for( Instance instance : InstanceHelpers.getAllInstances( application )) {
+
+			// Restore ALL the variables
+			List<ExportedVariable> variablesToRegenerate = new ArrayList<> ();
+			for( ExportedVariable var : instance.getComponent().exportedVariables.values()) {
+
+				// We only care about random ports
+				if( ! var.isRandom()
+						|| var.getRandomKind() != RandomKind.PORT )
+					continue;
+
+				// Find the overridden value.
+				// If it exists, restore it.
+				String value = instance.overriddenExports.get( var.getName());
+				if( value != null
+						&& ! acknowledgePort( application, instance, var.getName()))
+					variablesToRegenerate.add( var );
+			}
+
+			// Restoration may have failed for some. Regenerate new ports
+			for( ExportedVariable var : variablesToRegenerate ) {
+				this.logger.warning( "Generating a new random port for " + var.getName() + " in instance " + instance + " of " + application );
+				generateRandomPort( application, instance, var.getName());
+			}
+
+			// Save the updated model?
+			if( ! variablesToRegenerate.isEmpty())
+				ConfigurationUtils.saveInstances( application );
+		}
+	}
+
+
+	// Port Management
+
+	private static final int RABBITMQ_PORT = 35197;
+	private static final int PORT_MIN = 10000;
+	private static final int PORT_MAX = 65500;
+
```

Comment:
Indeed. To be managed as another issue.
We can add a list of forbidden ports in the DM's configuration (e.g. in the REST API or in its OSGi configuration).